lib/SecretsCrypto/secrets.c

```diff
@@ -0,0 +1,1000 @@
+ * Contact: Chris Adams <chris.adams@jollamobile.com>
+ * All rights reserved.
+ * BSD 3-Clause License, see LICENSE.
+ */
+
+#include "secrets.h"
+
+#include <glib.h>
+#include <gio/gio.h>
+#ifdef G_OS_UNIX
+#include <gio/gunixfdlist.h>
+#endif
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+/******************************* Data-Type Methods **************************/
+
+static struct Sailfish_Secrets_Result *Sailfish_Secrets_Result_new(
+		enum Sailfish_Secrets_Result_Code code,
+		int errorCode,
+		const char *errorMessage)
+{
+	struct Sailfish_Secrets_Result *result =
+	    (struct Sailfish_Secrets_Result *)
+	    malloc(sizeof(struct Sailfish_Secrets_Result));
+
+	result->code = code;
+	result->errorCode = errorCode;
+	result->errorMessage = errorMessage ? strndup(errorMessage, 512) : NULL;
+	result->refcount = 1;
+	return result;
+}
+
+void Sailfish_Secrets_Result_ref(struct Sailfish_Secrets_Result *result)
+{
+	if (result)
+		result->refcount = result->refcount + 1;
+}
+
+void Sailfish_Secrets_Result_unref(struct Sailfish_Secrets_Result *result)
+{
+	if (result) {
+		result->refcount = result->refcount - 1;
+		if (result->refcount == 0) {
+			free(result->errorMessage);
+			free(result);
+		}
+	}
+}
+
+struct Sailfish_Secrets_Secret_Identifier *
+Sailfish_Secrets_Secret_Identifier_new(
+		const char *name,
+		const char *collectionName)
+{
+	struct Sailfish_Secrets_Secret_Identifier *ident =
+	    (struct Sailfish_Secrets_Secret_Identifier *)
+	    malloc(sizeof(struct Sailfish_Secrets_Secret_Identifier));
+
+	ident->name = name ? strndup(name, 512) : NULL;
+	ident->collectionName = collectionName
+	    ? strndup(collectionName, 512) : NULL;
+	ident->refcount = 1;
+
+	return ident;
+}
+
+void Sailfish_Secrets_Secret_Identifier_ref(
+		struct Sailfish_Secrets_Secret_Identifier *ident)
+{
+	if (ident)
+		ident->refcount = ident->refcount + 1;
+}
+
+void Sailfish_Secrets_Secret_Identifier_unref(
+		struct Sailfish_Secrets_Secret_Identifier *ident)
+{
+	if (ident) {
+		ident->refcount = ident->refcount - 1;
+		if (ident->refcount == 0) {
+			free(ident->name);
+			free(ident->collectionName);
+			free(ident);
+		}
+	}
+}
+
+struct Sailfish_Secrets_Secret_FilterDatum *
+Sailfish_Secrets_Secret_FilterDatum_new(
+		const char *field,
+		const char *value)
+{
+	struct Sailfish_Secrets_Secret_FilterDatum *filter =
+	    (struct Sailfish_Secrets_Secret_FilterDatum *)
+	    malloc(sizeof(struct Sailfish_Secrets_Secret_FilterDatum));
+
+	filter->field = field ? strndup(field, 512) : NULL;
+	filter->value = value ? strndup(value, 512) : NULL;
+	filter->next = NULL;
+
+	return filter;
+}
+
+void Sailfish_Secrets_Secret_FilterDatum_unref(
+		struct Sailfish_Secrets_Secret_FilterDatum *filter)
+{
+	if (filter) {
+		struct Sailfish_Secrets_Secret_FilterDatum *curr = filter;
+		struct Sailfish_Secrets_Secret_FilterDatum *next = filter->next
+		    ? filter->next : NULL;
+
+		while (curr) {
```

Comment:
As I already mentioned I would suggest to use well-known frameworks to work with different kinds of linked lists.